src/Fuzz.elm

```diff
@@ -0,0 +1,381 @@
+module Fuzz exposing (Fuzzer, custom, unit, bool, order, array, char, float, floatRange, int, tuple, tuple3, tuple4, tuple5, result, string, percentage, map, maybe, intRange, list, frequency, frequencyOrCrash)
+
+{-| This is a library of `Fuzzer`s you can use to supply values to your fuzz tests.
+You can typically pick out which ones you need according to their types.
+
+A `Fuzzer a` knows how to create values of type `a`. It can create them
+randomly, and it can shrink them to more minimal values. Fuzzers can be
+filtered and mapped over.
+
+## Common Fuzzers
+@docs bool, int, intRange, float, floatRange, percentage, string, maybe, result, list, array
+
+## Working with Fuzzers
+@docs Fuzzer, map, frequency, frequencyOrCrash
+
+## Tuple Fuzzers
+Instead of using a tuple, consider using `fuzzN`.
+@docs tuple, tuple3, tuple4, tuple5
+
+## Uncommon Fuzzers
+@docs custom, char, unit, order
+
+-}
+
+import Array exposing (Array)
+import Char
+import Util exposing (..)
+import Shrink exposing (Shrinker)
+import Random exposing (Generator)
+import Random.Extra as Random
+import Fuzz.Internal as Internal
+
+
+{-| A Fuzzer is a
+[Random](http://package.elm-lang.org/packages/elm-lang/core/latest/Random)
+`Generator` paired with a shrinking strategy, or `Shrinker`. Shrinkers are defined
+in [`elm-community/shrink`](http://package.elm-lang.org/packages/elm-community/shrink/latest/).
+You will need to be familiar with both libraries to write custom fuzzers for your own types.
+Here is an example for a record:
+
+    type alias Position =
+        { x : Int, y : Int }
+
+
+    position : Fuzzer Position
+    position =
+        Fuzz.custom
+            (Random.map2 Position (Random.int 0 1919) (Random.int 0 1079))
+            (\{ x, y } -> Shrink.map Position (Shrink.int x) `Shrink.andMap` (Shrink.int y))
+
+Here is an example for a union type:
+
+    type Question
+        = Name String
+        | Age Int
+
+
+    question =
+        let
+            generator =
+                Random.bool `Random.andThen` (\b ->
+                    if b then
+                        Random.map Name string.generator
+                    else
+                        Random.map Age (Random.int 0 120)
+                 )
+
+            shrinker question =
+                case question of
+                    Name n ->
+                        Shrink.string n |> Shrink.map Name
+
+                    Age i ->
+                        Shrink.int i |> Shrink.map Age
+        in
+            Fuzz.custom generator shrinker
+-}
```

Comment:
I think this big long example goes with `custom`, not with the type.

I like to structure docs so that the relevant type comes first with a simple intuitive explanation. Details come out later in subsequent documentation. See [Maybe](http://package.elm-lang.org/packages/elm-lang/core/4.0.1/Maybe) or `Body` in [Http](http://package.elm-lang.org/packages/evancz/elm-http/3.0.1/Http).